File: materials/src/dce/Block.java
package dce;

import ir.IRInstruction;

import java.util.ArrayList;
import java.util.HashMap;

public class Block {

    ArrayList<IRInstruction> instructions = new ArrayList<>();

    public IRInstruction leader() {
        return instructions.get(0);
    }

    public IRInstruction branch() {
        return instructions.get(instructions.size() - 1);
    }

    int id = Block.globalId++;

    Block next1;
    Block next2;

    HashMap<String, IRInstruction> gen = new HashMap<>();
    HashMap<String, IRInstruction> kill = new HashMap<>();
    HashMap<String, IRInstruction> in = new HashMap<>();
    HashMap<String, IRInstruction> out = new HashMap<>();

    public boolean iterate() {
        HashMap<String, IRInstruction> temp = out;
        out = new HashMap<>();
        for (HashMap.Entry<String, IRInstruction> entry : in.entrySet()) {
            if (!kill.containsKey(entry.getKey())) {
                out.put(entry.getKey(), entry.getValue());
            }
        }
        out.putAll(gen);
        return out.equals(temp);
    }

    static int globalId = 0;

    public String toString() {
        return "Block_" + id + " [" + (next1 == null ? "" : next1.id) + " , " + (next2 == null ? "" : next2.id) + "]";
    }
}

File: materials/src/dce/Block.kt
package dce

import ir.IRInstruction

class Block {

    val instructions = ArrayList<IRInstruction>()
    val leader get() = instructions.first()
    val branch get() = instructions.last()

    val id = Block.id++

    var next1: Block? = null
    var next2: Block? = null

    companion object {
        var id = 0
    }

    override fun toString() = "Block_$id [${next1?.id ?: ""}, ${next2?.id ?: ""}]]"
}